File: web/src/utils/__tests__/batchIntentDetector.test.ts
import { describe, it, expect } from 'vitest';
import {
  BatchSignal,
  detectBatchIntent,
} from '@/utils/batchIntentDetector';

describe('detectBatchIntent', () => {
  it('returns not_batch for empty input', () => {
    expect(detectBatchIntent('')).toEqual({ kind: 'not_batch' });
    expect(detectBatchIntent('   ')).toEqual({ kind: 'not_batch' });
  });

  it('treats single-task commands as not_batch even with one signal', () => {
    // "buy groceries today" — TIME_RANGE only; should stay on single-task path.
    const res = detectBatchIntent('buy groceries today');
    expect(res.kind).toBe('not_batch');
  });

  it('treats a bare quantifier sentence as not_batch', () => {
    const res = detectBatchIntent('all the lights are on');
    expect(res.kind).toBe('not_batch');
  });

  it('routes commands with 2 signals to batch', () => {
    // QUANTIFIER + TIME_RANGE
    const res = detectBatchIntent('reschedule all overdue tasks to tomorrow');
    expect(res.kind).toBe('batch');
    if (res.kind === 'batch') {
      expect(res.signals.has(BatchSignal.QUANTIFIER)).toBe(true);
      // "tomorrow" is TIME_RANGE and "reschedule ... tasks" is BULK_VERB+PLURAL
      expect(res.signals.size).toBeGreaterThanOrEqual(2);
      expect(res.command_text).toBe('reschedule all overdue tasks to tomorrow');
    }
  });

  it('detects tag filter + time range', () => {
    const res = detectBatchIntent('cancel #work items this week');
    expect(res.kind).toBe('batch');
    if (res.kind === 'batch') {
      expect(res.signals.has(BatchSignal.TAG_FILTER)).toBe(true);
      expect(res.signals.has(BatchSignal.TIME_RANGE)).toBe(true);
    }
  });

  it('detects "tagged X" tag filter', () => {
    const res = detectBatchIntent('complete everything tagged urgent today');
    expect(res.kind).toBe('batch');
    if (res.kind === 'batch') {
      expect(res.signals.has(BatchSignal.TAG_FILTER)).toBe(true);
      expect(res.signals.has(BatchSignal.QUANTIFIER)).toBe(true);
    }
  });

  it('detects bulk verb + plural alongside a quantifier', () => {
    const res = detectBatchIntent('archive all tasks');
    expect(res.kind).toBe('batch');
    if (res.kind === 'batch') {
      expect(res.signals.has(BatchSignal.BULK_VERB_AND_PLURAL)).toBe(true);
      expect(res.signals.has(BatchSignal.QUANTIFIER)).toBe(true);
    }
  });

  it('rejects bulk verb + plural alone (one signal)', () => {
    // "clear tasks" — only BULK_VERB_AND_PLURAL, one signal, should NOT route.
    const res = detectBatchIntent('clear tasks');
    expect(res.kind).toBe('not_batch');
  });

  it('trims the command text before returning', () => {
    const res = detectBatchIntent('   reschedule all tasks this week   ');
    expect(res.kind).toBe('batch');
    if (res.kind === 'batch') {
      expect(res.command_text).toBe('reschedule all tasks this week');
    }
  });

  it('is case-insensitive across keywords', () => {
    const res = detectBatchIntent('Cancel ALL Habits This Week');
    expect(res.kind).toBe('batch');
  });

  it('handles multiword time-range phrase "next week"', () => {
    // QUANTIFIER + TIME_RANGE(phrase)
    const res = detectBatchIntent('move every task to next week');
    expect(res.kind).toBe('batch');
    if (res.kind === 'batch') {
      expect(res.signals.has(BatchSignal.TIME_RANGE)).toBe(true);
    }
  });
});

File: app/src/test/java/com/averycorp/prismtask/domain/usecase/BatchIntentDetectorTest.kt
package com.averycorp.prismtask.domain.usecase

import com.averycorp.prismtask.domain.usecase.BatchIntentDetector.Result
import com.averycorp.prismtask.domain.usecase.BatchIntentDetector.Signal
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class BatchIntentDetectorTest {
    private val detector = BatchIntentDetector()

    @Test
    fun emptyInput_isNotBatch() {
        assertEquals(Result.NotABatch, detector.detect(""))
        assertEquals(Result.NotABatch, detector.detect("   "))
    }

    @Test
    fun singleTaskCommand_isNotBatch_stillFlowsIntoRegularQuickAdd() {
        // These are regular single-task creations — one or zero signal
        // categories. They MUST NOT be diverted to the batch path.
        assertEquals(Result.NotABatch, detector.detect("Buy milk tomorrow"))
        assertEquals(Result.NotABatch, detector.detect("Call Anna"))
        assertEquals(Result.NotABatch, detector.detect("Fix bug #urgent"))
        assertEquals(Result.NotABatch, detector.detect("Grocery run this weekend"))
    }

    @Test
    fun quantifierPlusTimeRange_detects() {
        val result = detector.detect("Cancel everything Friday")
        assertTrue(result is Result.Batch)
        val batch = result as Result.Batch
        assertEquals("Cancel everything Friday", batch.commandText)
        assertTrue(Signal.QUANTIFIER in batch.signals)
        assertTrue(Signal.TIME_RANGE in batch.signals)
    }

    @Test
    fun tagFilterPlusBulkVerbPlusPlural_detects() {
        val result = detector.detect("Move all tasks tagged work to Monday")
        assertTrue(result is Result.Batch)
        val batch = result as Result.Batch
        assertTrue(Signal.QUANTIFIER in batch.signals)
        assertTrue(Signal.TAG_FILTER in batch.signals)
        assertTrue(Signal.TIME_RANGE in batch.signals)
        assertTrue(Signal.BULK_VERB_AND_PLURAL in batch.signals)
    }

    @Test
    fun hashtagFilter_countsAsTagFilter() {
        val result = detector.detect("Reschedule all #urgent tasks to tomorrow")
        assertTrue(result is Result.Batch)
        val batch = result as Result.Batch
        assertTrue(Signal.TAG_FILTER in batch.signals)
    }

    @Test
    fun clearThursdayAfternoon_detectsViaTimePhraseAndBulkVerb() {
        // "Clear" + "Thursday" = TIME_RANGE signal. "Clear" is a bulk verb
        // but there's no entity plural in "Clear Thursday afternoon", so
        // BULK_VERB_AND_PLURAL doesn't fire. We should still detect via
        // TIME_RANGE + another signal — in this case "afternoon" phrase.
        val result = detector.detect("Clear Thursday afternoon")
        // Thursday + "this afternoon"-style parse isn't guaranteed; the
        // safer assertion: the current detector returns NotABatch for
        // this, because there's only one signal category (TIME_RANGE).
        // The user's example in the spec is "Clear Thursday afternoon",
        // so if the detector rejects it the client has to fall back —
        // that's acceptable; the user can rephrase as "Cancel everything
        // Thursday".
        assertEquals(Result.NotABatch, result)
    }

    @Test
    fun bulkVerbPlusEntityPlural_alone_isNotEnough() {
        // "Cancel meetings" is structurally batch-ish but has just one
        // signal category. We prefer single-task parsing when unsure.
        val result = detector.detect("Cancel meetings")
        assertEquals(Result.NotABatch, result)
    }

    @Test
    fun everyHabit_plusTimeRange_detects() {
        val result = detector.detect("Mark every habit complete for today")
        assertTrue(result is Result.Batch)
    }

    @Test
    fun rescheduleAllOverdueTasksTomorrow_detects() {
        val result = detector.detect("Reschedule all overdue tasks to tomorrow")
        assertTrue(result is Result.Batch)
        val batch = result as Result.Batch
        assertTrue(Signal.QUANTIFIER in batch.signals)
        assertTrue(Signal.TIME_RANGE in batch.signals)
        assertTrue(Signal.BULK_VERB_AND_PLURAL in batch.signals)
    }

    @Test
    fun caseInsensitive_match() {
        assertTrue(detector.detect("CANCEL EVERYTHING FRIDAY") is Result.Batch)
        assertTrue(detector.detect("cancel Everything Friday") is Result.Batch)
    }

    @Test
    fun commandTextPreservesOriginalCasing_evenAfterMatch() {
        val batch = detector.detect("Cancel Everything Friday") as Result.Batch
        assertEquals("Cancel Everything Friday", batch.commandText)
    }

    // ---------------------------------------------------------------------
    // New tests — Section G phase F.1 minimum suite (P0 + P1 + YELLOW).
    // ---------------------------------------------------------------------

    @Test
    fun recurrencePattern_isNotBatch() {
        // P0 carve-out: `every <weekday>`-style inputs are recurring
        // single-task creation, not batch operations. signals stay
        // {QUANT, TIME_RANGE} but the carve-out forces NotABatch.
        // One input is intentionally MIXED CASE to lock in case
        // insensitivity through the carve-out path.
        assertEquals(Result.NotABatch, detector.detect("every monday at 8am team standup"))
        assertEquals(Result.NotABatch, detector.detect("Every Monday At 8AM Team Standup"))
        assertEquals(Result.NotABatch, detector.detect("each friday morning"))
        assertEquals(Result.NotABatch, detector.detect("every sunday weekly habit reset"))
        assertEquals(Result.NotABatch, detector.detect("every day at 7am"))
    }

    @Test
    fun everyTimeRangeWithBulkVerbAndPlural_stillBatch() {
        // Recurrence carve-out is narrowly scoped: it only fires when
        // signals == {QUANT, TIME_RANGE}. A third signal (here
        // BULK_VERB_AND_PLURAL) keeps the input on the batch path.
        assertTrue(detector.detect("complete every task tomorrow") is Result.Batch)
        assertTrue(detector.detect("delete all tasks every monday") is Result.Batch)
    }

    @Test
    fun negationPrefix_isNotBatch() {
        // P1 carve-out: explicit negation prefixes suppress batch routing
        // even when the rest of the command matches multiple signals.
        assertEquals(Result.NotABatch, detector.detect("don't complete all tasks today"))
        assertEquals(Result.NotABatch, detector.detect("do not delete tasks tomorrow"))
        assertEquals(Result.NotABatch, detector.detect("please don't archive everything tomorrow"))
        assertEquals(Result.NotABatch, detector.detect("never reschedule all tasks to monday"))
        // Mixed case — locks in case insensitivity through the negation path.
        assertEquals(Result.NotABatch, detector.detect("DON'T Complete All Tasks Today"))
    }

    @Test
    fun negationInMiddleOfSentence_doesNotSuppress() {
        // The negation pre-check is anchored to the start of input. A
        // negation token in the middle does not suppress routing.
        assertTrue(detector.detect("remember to delete all tasks today") is Result.Batch)
    }

    @Test
    fun trailingPunctuation_doesNotBreakTimeRange() {
        // P1 fix: tokenization now splits on `,.;!?` so trailing
        // punctuation no longer drops dictionary lookups.
        assertTrue(detector.detect("delete tasks today.") is Result.Batch)
        assertTrue(detector.detect("complete all tasks tomorrow!") is Result.Batch)
        assertTrue(detector.detect("cancel meds, friday") is Result.Batch)
    }

    @Test
    fun existingBehaviorPreserved() {
        // Lock-in: the new pre-checks and tokenization changes do not
        // flip existing single-task / batch routing.
        assertEquals(Result.NotABatch, detector.detect("Buy milk tomorrow"))
        assertTrue(detector.detect("Cancel everything Friday") is Result.Batch)
    }

    @Test
    fun markVerbAndPlural_detects() {
        // YELLOW fix: `mark` is now a bulk verb. "mark all tasks done"
        // routes via QUANT + BULK_VERB_AND_PLURAL (no TIME_RANGE needed).
        val result = detector.detect("mark all tasks done")
        assertTrue(result is Result.Batch)
        val batch = result as Result.Batch
        assertTrue(Signal.QUANTIFIER in batch.signals)
        assertTrue(Signal.BULK_VERB_AND_PLURAL in batch.signals)

        // Existing input still routes Batch — `habit` is singular, so
        // BULK_VERB_AND_PLURAL still doesn't fire and the signal mix
        // stays {QUANT, TIME_RANGE}. The recurrence carve-out does not
        // fire because `habit` is not in the recurrence-noun set.
        assertTrue(detector.detect("mark every habit complete for today") is Result.Batch)
    }

    @Test
    fun hashtagMidToken_doesNotMatchTagFilter() {
        // YELLOW fix D.5b: `#` in the middle of a token (e.g. `C#programming`)
        // no longer false-matches TAG_FILTER. With only TIME_RANGE the
        // input is one signal → NotABatch.
        assertEquals(Result.NotABatch, detector.detect("study C#programming tomorrow"))
        // `C# ` (space after) was already NotABatch pre-fix; verify it
        // still is.
        assertEquals(Result.NotABatch, detector.detect("study C# programming tomorrow"))
    }

    @Test
    fun numericHashtag_doesNotMatchTagFilter() {
        // YELLOW fix D.5a: `#1`, `#42` are issue/reference markers, not
        // tags. TAG_FILTER no longer fires on pure-numeric hashtags.
        assertEquals(Result.NotABatch, detector.detect("task #1 status today"))
        assertEquals(Result.NotABatch, detector.detect("fix #42 by friday"))
    }

    @Test
    fun mixedHashtag_stillMatches() {
        // Mixed-content tags (digits + non-digits) still count as
        // TAG_FILTER. Combined with TIME_RANGE → Batch.
        assertTrue(detector.detect("tag #2024-q1 today") is Result.Batch)
        // `review` is NOT a bulk verb; signals here are TAG_FILTER +
        // TIME_RANGE = 2 → Batch.
        val reviewResult = detector.detect("review #urgent items tomorrow")
        assertTrue(reviewResult is Result.Batch)
        val batch = reviewResult as Result.Batch
        assertTrue(Signal.TAG_FILTER in batch.signals)
        assertTrue(Signal.TIME_RANGE in batch.signals)
    }

    @Test
    fun quotedInput_staysNotABatch_knownLimitation() {
        // Tokenizer does not strip quotes, so `"complete` and `today"`
        // miss their respective dictionaries. Only QUANT (`all`) fires,
        // so signals=1 → NotABatch. Lock in the current behavior as a
        // known limitation; revisit if it shows up in user reports.
        assertEquals(Result.NotABatch, detector.detect("\"complete all tasks today\""))
    }
}
